### PR TITLE
[LLVM][DWARF] Create thread safe context for DWP DWARFContext

### PR DESCRIPTION
**Pull request by ayermolo**
Right now DWARFContext that is created is the same when input is DWO or DWP file.
Changed so that for DWP it creates a thread safe context. This allows for a multi threaded safe access to {cu,tu}-index

### CODE REVIEW

**Comment by ayermolo** [comment]
@dwblaikie @clayborg ping :D

File: llvm/lib/DebugInfo/DWARF/DWARFContext.cpp [hunk 1]
```diff
@@ -575,8 +575,12 @@ class ThreadUnsafeDWARFContextState : public DWARFContext::DWARFContextState {
...
     auto S = std::make_shared<DWOFile>();
     S->File = std::move(Obj.get());
-    S->Context = DWARFContext::create(*S->File.getBinary(),
-                                      DWARFContext::ProcessDebugRelocations::Ignore);
+    StringRef FileName = S->File.getBinary()->getFileName();
+    S->Context = DWARFContext::create(
+        *S->File.getBinary(), DWARFContext::ProcessDebugRelocations::Ignore,
+        nullptr, "", WithColor::defaultErrorHandler,
+        WithColor::defaultWarningHandler,
+        FileName.find(".dwp") != StringRef::npos);
```
Comment (clayborg) [inline llvm/lib/DebugInfo/DWARF/DWARFContext.cpp:583]: Add a comment stating why your are enabling threading support and make a local variable so we can add a comment:
```
// Allow multi-threaded access if there is a .dwp file as the CU index 
// and TU index might be accessed from multiple threads.
bool ThreadSafe = FileName.find(".dwp") != StringRef::npos;
S->Context = DWARFContext::create(
        *S->File.getBinary(), DWARFContext::ProcessDebugRelocations::Ignore,
        nullptr, "", WithColor::defaultErrorHandler,
        WithColor::defaultWarningHandler, ThreadSafe);
```

**Review by clayborg — CHANGES_REQUESTED** [review]

**Review by clayborg — APPROVED** [review]

**Comment by dwblaikie** [comment]
Not sure I follow - wouldn't it be suitable for a ThreadUnsafeDWARFContext to create a ThreadUnsafeDWARFContext for the DWP? And only a ThreadSafeDWARFContext would create a ThreadSafeDWARFContext for the DWP?

**Comment by clayborg** [comment]
> Not sure I follow - wouldn't it be suitable for a ThreadUnsafeDWARFContext to create a ThreadUnsafeDWARFContext for the DWP? And only a ThreadSafeDWARFContext would create a ThreadSafeDWARFContext for the DWP?

No one is creating a thread safe context right now except llvm-gsymutil. So we could enable thread safe everywhere, but people seemed to think there would be some performance hit if that was done.

**Comment by clayborg** [comment]
> > Not sure I follow - wouldn't it be suitable for a ThreadUnsafeDWARFContext to create a ThreadUnsafeDWARFContext for the DWP? And only a ThreadSafeDWARFContext would create a ThreadSafeDWARFContext for the DWP?
> 
> No one is creating a thread safe context right now except llvm-gsymutil. So we could enable thread safe everywhere, but people seemed to think there would be some performance hit if that was done.

If this issue here was we had a thread safe DWARContext for the main DWARF, and this wasn't being correctly propagated into the .dwp DWARFContext, then this should be fixed by detecting the thread safe setting from the DWARContext and using that setting for any child DWARFContext (.dwp and any .dwo files that are not in a .dwp file) objects.

**Comment by dwblaikie** [comment]
> > > Not sure I follow - wouldn't it be suitable for a ThreadUnsafeDWARFContext to create a ThreadUnsafeDWARFContext for the DWP? And only a ThreadSafeDWARFContext would create a ThreadSafeDWARFContext for the DWP?
> > 
> > 
> > No one is creating a thread safe context right now except llvm-gsymutil. So we could enable thread safe everywhere, but people seemed to think there would be some performance hit if that was done.
> 
> If this issue here was we had a thread safe DWARContext for the main DWARF, and this wasn't being correctly propagated into the .dwp DWARFContext, then this should be fixed by detecting the thread safe setting from the DWARContext and using that setting for any child DWARFContext (.dwp and any .dwo files that are not in a .dwp file) objects.

But this code change is being made in the `ThreadUnsafeDWARFContextState` - the thread unsafe class is making a thread safe DWP context. That doesn't seem right? Shouldn't only a ThreadSafeDWARFContextState create a thread safe context for the DWP?

**Comment by clayborg** [comment]
So sounds like we need to change this patch to propagate the "ThreadSafe" from one context to any child DWARFContext things (like for owned .dwp and .dwo contexts. I would recommend adding a new virtual function to DWARFContext:
```
virtual bool IsThreadSafe() = 0;
```
Then have ThreadUnsafeDWARFContextState override this and return false, and ThreadSafeState override it return true. Then it will be easy to propagate the setting using the current "DWARFContext::State" iivar from a DWARFContext to create another.

**Comment by ayermolo** [comment]
OK, will make the change so that DWO/DWP DWARFContext is controlled by main binary DWARFContext thread safetiness.

**Review by dwblaikie — APPROVED** [review]
Sounds OK

**Pull request merged**